pkg/kubelet/userns_manager.go

```diff
@@ -0,0 +1,476 @@
+	return (b.data[index/bitsUint32]>>(index%bitsUint32))&0x1 == 1
+}
+
+func (b *bitArray) findAndSetFirstZero() (uint32, error) {
+	for i := b.firstIndex; i < len(b.data); i++ {
+		if b.data[i] == 0xFFFFFFFF {
+			continue
+		}
+		for j := uint32(0); j < bitsUint32; j++ {
+			if (b.data[i]>>j)&0x1 == 0 {
+				v := uint32(i)*bitsUint32 + j
+				b.set(v)
+				b.firstIndex = int(i)
+				return v, nil
+			}
+		}
+	}
+	return 0, fmt.Errorf("could not find an empty slot")
+}
+
+func (b *bitArray) clear(index uint32) {
+	i := index / bitsUint32
+	if i < uint32(b.firstIndex) {
+		b.firstIndex = int(i)
+	}
+	b.data[i] &= ^(1 << (index % bitsUint32))
+}
+
+// length for the user namespace to create (65536).
+const userNsLength = math.MaxUint16 + 1
+
+// Limit the total number of pods using userns in this node to this value.
+// This is an alpha limitation that will probably be lifted later.
+const maxPods = 1024
+
+type usernsManager struct {
+	used     *bitArray
+	usedBy   map[string]uint32
+	removed  int
+	currPods int
+	kl       *Kubelet
+	sync.Mutex
+}
+
+// UserNamespace holds the configuration for the user namespace.
+type userNamespace struct {
+	// UIDs mappings for the user namespace.
+	UIDMappings []idMapping `json:"uidMappings"`
+	// GIDs mappings for the user namespace.
+	GIDMappings []idMapping `json:"gidMappings"`
+}
+
+// Pod user namespace mapping
+type idMapping struct {
+	// Required.
+	HostId uint32 `json:"hostId"`
+	// Required.
+	ContainerId uint32 `json:"containerId"`
+	// Required.
+	Length uint32 `json:"length"`
+}
+
+func MakeUserNsManager(kl *Kubelet) (*usernsManager, error) {
+	m := usernsManager{
+		// Create a bitArray for all the UID space (2^32).
+		// As a by product of that, no index param to bitArray can be out of bounds (index is uint32).
+		used:   makeBitArray((math.MaxUint32 + 1) / userNsLength),
+		usedBy: make(map[string]uint32),
+		kl:     kl,
+	}
+	// First block is reserved for the host.
+	m.used.set(0)
+
+	// Second block will be used for phase II. Don't assign that range for now.
+	m.used.set(1)
+
+	if kl != nil {
+		found, err := kl.listPodsFromDisk()
+		if err != nil {
+			return nil, fmt.Errorf("user namespace manager can't read pods from disk: %w", err)
+		}
+
+		for _, uid := range found {
+			if err := m.recordPodMappings(uid); err != nil {
+				return nil, err
+			}
+		}
+	}
+	return &m, nil
+}
+
+// recordPodMappings registers the range used for the user namespace iif the
+// usernsConfFile exits in the pod directory.
+func (m *usernsManager) recordPodMappings(pod types.UID) error {
+	usernsConfFile := m.getUserNamespaceMappingsFile(pod)
+
+	// If the size is too big, something is fishy. Don't read it all in mem.
+	// Ignore errors, the file might not be created yet and afterwards we
+	// check which error was returned.
+	if info, err := os.Stat(usernsConfFile); err == nil && info.Size() >= 1024*1024 {
+		return fmt.Errorf("pod user namespace file is too big: %q", usernsConfFile)
```

Comment:
Can we provide some context about the path to the file that's too big? Any other contextual data we can provide for observability?

Reply:
Thanks for spotting it.  Thinking more of it, I am not even sure why we left this check.  I'll drop it

Reply:
@giuseppe I'd keep it, see: https://github.com/kubernetes/kubernetes/pull/111090/files#r919934584

@sftim context about the path to the file? What do you mean? We are showing the path there, that file is not supposed to be touched by anyone, this check just makes sure that in the unlikely case that a bad actor touched it, we won't allocate a huge amount of RAM to read it.

Reply:
If I saw this error in my cluster logs, I would have to read the source code to understand what was going on and what a might be appropriate as a fix. Ideally, the logs would give the cluster operator more of a clue.

If this never gets logged, well, OK, it doesn't have to be so clear.

Aside: `usernsConfFile` is not really the path to a _configuration_ file as I'd use the term. Maybe call it `userNamespaceMappingsFilePath`?

Reply:
@sftim Yes, this should really never happen. You would only see this if you modified the file kubelet private file to be bigger manually, or we have a very nasty bug when writing to it.